library/std/src/ffi/os_str.rs

```diff
@@ -582,15 +582,21 @@ impl OsString {
     #[unstable(feature = "os_string_truncate", issue = "133262")]
     pub fn truncate(&mut self, len: usize) {
         self.as_os_str().inner.check_public_boundary(len);
-        self.inner.truncate(len);
+        // SAFETY: The length was just checked to be at a valid boundary.
+        unsafe { self.inner.truncate_unchecked(len) };
     }
 
-    /// Provides plumbing to core `Vec::extend_from_slice`.
-    /// More well behaving alternative to allowing outer types
-    /// full mutable access to the core `Vec`.
+    /// Provides plumbing to `Vec::extend_from_slice` without giving full
+    /// mutable access to the `Vec`.
+    ///
+    /// # Safety
+    ///
+    /// Extending the buffer with this slice must not require encoding-dependent
+    /// surrogate joining.
```

Comment:
"must not require encoding-dependent surrogate joining" is deferring the safety requirement to instead be "you must know exactly how the implementation works", so I think this is not a significant improvement on messaging for the (ahem) modularity of reasoning.

unless we can concisely describe what requires that, in which case "don't put that in" should be the requirement instead.

and I am pretty sure we can?

Reply:
specifically "the last bytes of `self` must not encode a leading/high surrogate for a surrogate pair, and the first bytes of `other` must not encode a trailing/low surrogate of a surrogate pair". is there another case?